Add missing FillOp to winograd lowering

RoboTux opened the pull request:
Winograd lowering involves a number of matmul and batch_matmul which
are currently passed tensor.empty result as out parameter, thereby
are undefined behaviour. This commit adds the necessary linalg.fill.

RoboTux committed:
[Linalg] Add missing FillOp to winograd lowering

Winograd lowering involves a number of matmul and batch_matmul which
are currently passed tensor.empty result as out parameter, thereby
are undefined behaviour. This commit adds the necessary linalg.fill.

Max191 commented on mlir/test/Dialect/Linalg/transform-tile-and-winograd-rewrite.mlir:
```diff
@@ -44,9 +44,9 @@ module attributes {transform.with_named_sequence} {
-// CHECK:      %[[S11:.*]] = linalg.matmul
-// CHECK:      %[[S13:.*]] = linalg.matmul
-// CHECK:      %[[INSERTED_SLICE:.*]] = tensor.insert_slice %[[S13]] into %[[ARG6]][0, 0, %[[ARG5]], %[[ARG3]]] [6, 6, 1, 1] [1, 1, 1, 1]
+// CHECK:      %[[S12:.*]] = linalg.matmul
+// CHECK:      %[[S15:.*]] = linalg.matmul
+// CHECK:      %[[INSERTED_SLICE:.*]] = tensor.insert_slice %[[S15]] into %[[ARG6]][0, 0, %[[ARG5]], %[[ARG3]]] [6, 6, 1, 1] [1, 1, 1, 1]
```
Can you also check that the `outs` has the result of the fill?

Max191 commented on mlir/test/Dialect/Linalg/transform-tile-and-winograd-rewrite.mlir:
```diff
@@ -56,20 +56,20 @@ module attributes {transform.with_named_sequence} {
-// CHECK:          %[[INSERTED_SLICE_9:.*]] = tensor.insert_slice %[[S17]] into %[[ARG10]][0, 0, 0, 0, %[[ARG7]], %[[ARG9]]] [6, 6, 1, 1, 1, 1] [1, 1, 1, 1, 1, 1]
-// CHECK:          scf.yield %[[INSERTED_SLICE_9]]
+// CHECK:          %[[EXTRACTED_SLICE_9:.*]] = tensor.extract_slice %[[EXTRACTED_SLICE]][%[[ARG7]], 0, 0, %[[ARG9]]] [1, 6, 6, 1] [1, 1, 1, 1]
+// CHECK:          %[[S16:.*]] = linalg.matmul
+// CHECK:          %[[S19:.*]] = linalg.matmul
```
Same here, check that fill is in the `outs`

Max191 commented on mlir/test/Dialect/Linalg/transform-tile-and-winograd-rewrite.mlir:
```diff
@@ -56,20 +56,20 @@ module attributes {transform.with_named_sequence} {
 // CHECK:      %[[EXTRACTED_SLICE:.*]] = tensor.extract_slice %[[ARG0]][0, %[[S10]], %[[S11]], 0] [2, 6, 6, 5] [1, 1, 1, 1]
-// CHECK:      %[[EXTRACTED_SLICE_7:.*]] = tensor.extract_slice %[[S2]][0, 0, %[[ARG3]], %[[ARG5]], 0, 0] [6, 6, 1, 1, 2, 5] [1, 1, 1, 1, 1, 1]
-// CHECK:      %[[S12:.*]] = scf.for %[[ARG7:.*]] = %[[C0]] to %[[C2]] step %[[C1]] iter_args(%[[ARG8:.*]] = %[[EXTRACTED_SLICE_7]])
+// CHECK:      %[[EXTRACTED_SLICE_8:.*]] = tensor.extract_slice %[[S2]][0, 0, %[[ARG3]], %[[ARG5]], 0, 0] [6, 6, 1, 1, 2, 5] [1, 1, 1, 1, 1, 1]
```
Why are the names of all these checks changing? It doesn't seem to be necessary.

Max191 reviewed (changes requested):
Thanks for catching this. It's a little hard to see what's changing in the lit tests, since all the names in the checks have changed. Could you keep the names there the same? Also, please add checks in the decomposition tests that fill ops are created and used in the matmuls.

RoboTux committed:
Clean up tests

Reduce diff noise in tests and add tensor.empty and linalg.fill check in
decomposition tests.

RoboTux commented:
> Thanks for catching this. It's a little hard to see what's changing in the lit tests, since all the names in the checks have changed. Could you keep the names there the same? Also, please add checks in the decomposition tests that fill ops are created and used in the matmuls.

I've reduced the noise somewhat but adding the fill inevitably resequence quite a bit of SSA values. It also leads to some FileCheck variable reuse because not everything was renamed. Alternatively if you prefer I could do a NFC patch that just update all FileCheck variable like in my first attempt but doesn't have the fix. Then in a second patch I can add just the fix and the new CHECK and land that as a rebase rather than squash if that's possible in github.

Max191 commented on mlir/test/Dialect/Linalg/transform-tile-and-winograd-rewrite.mlir:
```diff
@@ -44,9 +51,13 @@ module attributes {transform.with_named_sequence} {
+// CHECK:      %[[S10:.*]] = tensor.empty() : tensor<6x3xf32>
+// CHECK:      %[[S11:.*]] = linalg.fill ins(%[[CST_6]] : f32) outs(%[[S10]] : tensor<6x3xf32>) -> tensor<6x3xf32>
+// CHECK:      %[[S12:.*]] = linalg.matmul ins(%[[CST_5]], %[[EXTRACTED_SLICE]] : tensor<6x3xf32>, tensor<3x3xf32>) outs(%[[S11]] : tensor<6x3xf32>) -> tensor<6x3xf32>
+// CHECK:      %[[S13:.*]] = tensor.empty() : tensor<6x6xf32>
+// CHECK:      %[[S14:.*]] = linalg.fill ins(%[[CST_6]] : f32) outs(%[[S13]] : tensor<6x6xf32>) -> tensor<6x6xf32>
+// CHECK:      %[[S15:.*]] = linalg.matmul ins(%[[S12]], %[[CST_4]] : tensor<6x3xf32>, tensor<3x6xf32>) outs(%14 : tensor<6x6xf32>) -> tensor<6x6xf32>
```
```suggestion
// CHECK:      %[[S13:.*]] = tensor.empty() : tensor<6x6xf32>
// CHECK:      %[[S14:.*]] = linalg.fill ins(%[[CST_6]] : f32) outs(%[[S13]] : tensor<6x6xf32>) -> tensor<6x6xf32>
// CHECK:      %[[S15:.*]] = linalg.matmul ins(%[[S12]], %[[CST_4]] : tensor<6x3xf32>, tensor<3x6xf32>) outs(%[[S14]] : tensor<6x6xf32>) -> tensor<6x6xf32>
```

Max191 reviewed (approved):
Thanks! This looks much better.

Max191 commented:
> I've reduced the noise somewhat but adding the fill inevitably resequence quite a bit of SSA values. It also leads to some FileCheck variable reuse because not everything was renamed. Alternatively if you prefer I could do a NFC patch that just update all FileCheck variable like in my first attempt but doesn't have the fix. Then in a second patch I can add just the fix and the new CHECK and land that as a rebase rather than squash if that's possible in github.

It looks better now, so no worries. It is usually good to have more meaningful FileCheck variable names, but I probably should have caught that on the original PRs. It makes lit tests more readable and maintainable when the variables aren't just numbered ssa values (you don't need to do anything about that in this PR, just noting).

RoboTux committed:
Fix use of hardcoded SSA value

Co-authored-by: Max191 <44243577+Max191@users.noreply.github.com>
